### PR TITLE
add TORCH_FORCE_SYNCHRONOUS_COLLECTIVES to force functional collectives to be synchronous

### PR DESCRIPTION
**Pull request by bdhirsh**
One way that OOM's can show up with using torch.compile with distributed code is if our compile fails to issue a `wait_tensor` properly on a functional collective.

This PR adds an env var that forces all functional collectives to be synchronous at runtime (making issuing wait_tensor at runtime a no-op, so if we fail to issue wait_tensors properly at runtime then we will not leak memory).

If we see a mem leak / OOM when using compile with distributed code, this flag should at least quickly tell us if the mem leak is coming from not waiting on a collective, vs. something else entirely.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @yifuwang / @wconstab, wdyt of this env var?

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128331
* #128045
* #130338

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by bdhirsh** [comment]
yep, very fair (@drisspg - per that BE project you were looking into, do we have a nice way for users to find TORCH_* env vars already? Otherwise maybe if/once we agree that this env var is reasonable, I can look for a place in the docs to mention it)

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 1]
```diff
@@ -10,11 +10,25 @@
 
 namespace {
 
+// Forces all functional collectives to wait immediately.
+// Useful for debugging memory leaks with compile that surface
+// due to compile not properly waiting on functional collectives.
+bool force_synchronous_functional_collectives() {
+  static char const* temp = getenv("TORCH_FORCE_SYNCHRONOUS_COLLECTIVES");
```
Comment (Skylion007) [inline torch/csrc/distributed/c10d/Functional.cpp:17]: Use std::getenv. Also what if the user wants to explicitly disable it?

**Comment by wconstab** [comment]
I like the idea of adding the env, but not sure how to name it.

most of the distributed envs are named `TORCH_NCCL_*` and @shuqiang recently doc'd them all in [this page](https://github.com/pytorch/pytorch/pull/128235), you should probably add an entry there and also in compile docs?

The 2 confusing points regarding naming and behavior are 
 - its not NCCL specific, so probably TORCH_ is better
 - the env only affects functional collectives, which might confuse users of non-functional collectives

TORCH_FUNCOL_SYNC?

**Comment by bdhirsh** [comment]
> The 2 confusing points regarding naming and behavior are

@wconstab thanks for the feedback. The point about most env vars including `NCCL` in their name but this not being nccl-specific is fair. `TORCH_FUNCOL_SYNC` sounds reasonable to me, I can change it to that.

File: torch/csrc/distributed/c10d/Functional.cpp (hunk 1 above)
Comment (bdhirsh) [reply torch/csrc/distributed/c10d/Functional.cpp:17]: I'll probably just switch to use the same API's to read envvars as the other distributed code: https://github.com/pytorch/pytorch/blob/main/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L780

**Comment by yifuwang** [comment]
This looks great! Thanks for adding it.

Curious what are the most common sources of unwaited collectives based on your observation. It is more due to graph break boundary handling, or users calling the `_c10d_functional` ops without calling wait? For the latter, we can add a pass to lint/fix the program.

**Review by kwen2501 — APPROVED** [review]

**Pull request closed without merging**